refactor: Make CTxMemPoolEntry only explicitly copyable

sedited opened the pull request:
This has the goal of prohibiting users from accidentally creating runtime failures, e.g. by interacting with iterator_to with a copied entry. This was brought up here:  https://github.com/bitcoin/bitcoin/pull/28886#issuecomment-1814794954.

CTxMemPoolEntry is already implicitly not move-constructable. So be explicit about this and use a std::list to collect the values in the policy_estimator fuzz test instead of a std::vector.

ajtowns commented on src/test/fuzz/policy_estimator.cpp:
```diff
@@ -58,7 +58,16 @@ FUZZ_TARGET(policy_estimator, .init = initialize_policy_estimator)
+                        entry.GetSequence(),
+                        entry.GetSpendsCoinbase(),
+                        entry.GetSigOpCost(),
+                        entry.GetLockPoints());
```
Replacing a simple copy with listing all the details of the object seems cumbersome. If we want to just prevent accidental copies, how about doing something like this instead:

```c++
class CTxMemPoolEntry {
private:
    CTxMemPoolEntry(const CTxMemPoolEntry&) = default;
    struct ExplicitCopyTag { explicit ExplicitCopyTag() = default; };
public:
    static constexpr ExplicitCopyTag ExplicitCopy;
    CTxMemPoolEntry(ExplicitCopyTag /*unused*/, const CTxMemPoolEntry& entry) : CTxMemPoolEntry(entry) { }
    ...
};

    mempool_entries.emplace_back(CTxMemPoolEntry::ExplicitCopy, entry);
```

TheCharlatan replied:
This looks nice, I'll try to apply it.

sedited committed:
refactor: Make CTxMemPoolEntry only explicitly copyable

This has the goal of prohibiting users from accidentally creating
runtime failures, e.g. by interacting with iterator_to with a copied
entry.

CTxMemPoolEntry is already implicitly not move-constructable. So be
explicit about this and use a std::list to collect the values in the
policy_estimator fuzz test instead of a std::vector.

Co-authored-by: Anthony Towns <aj@erisian.com.au>

TheCharlatan commented:
Thank you @ajtowns,

Updated b9e4b4cb275e998376b0e3e68b6c22c5b1877fbb -> b8b624b36ebeb167da2f9c6d12f563f6849ccd60 ([CTxMemPoolEntryNonCopy_0](https://github.com/TheCharlatan/bitcoin/tree/CTxMemPoolEntryNonCopy_0) -> [CTxMemPoolEntryNonCopy_1](https://github.com/TheCharlatan/bitcoin/tree/CTxMemPoolEntryNonCopy_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/CTxMemPoolEntryNonCopy_0..CTxMemPoolEntryNonCopy_1))
* Applied @ajtowns' [suggestion](https://github.com/bitcoin/bitcoin/pull/28903#discussion_r1397661685), making `CTxMemPoolEntry` only copy-able if a special tag is provided.

TheCharlatan commented:
Thanks for the review @maflcko 

Re https://github.com/bitcoin/bitcoin/pull/28903#issuecomment-1818810741

> For reference, the move constructor is already deleted, according to a compiler:

Removed the explicit move constructor deletion again.

TheCharlatan commented:
Reverted the last change again.

ajtowns commented on src/kernel/mempool_entry.h:
```diff
@@ -71,6 +71,11 @@ class CTxMemPoolEntry
     typedef std::set<CTxMemPoolEntryRef, CompareIteratorByHash> Children;
 
 private:
+    CTxMemPoolEntry(const CTxMemPoolEntry&) = default;
```
Might be good to add a comment explaining why copying this is often a bad thing and should only be done explicitly? (I think it's because `m_parents` and `m_children` need to stay in sync with the container holding this entry -- if you leave the copy out of the container, those sets will become bad; if you put it back into the container, those sets will probably be wrong)

TheCharlatan replied:
I'll do this in #28886.

ajtowns reviewed (commented):
ACK 705e3f1de00bf30d728addd52a790a139d948e32

ismaelsadeeq reviewed (approved):
ACK 705e3f1de00bf30d728addd52a790a139d948e32

achow101 commented:
ACK 705e3f1de00bf30d728addd52a790a139d948e32
